net: Allow -proxy=[::1] on nodes with IPV6 lo only

m3dwards opened the pull request:
This is similar to (but does not fix) https://github.com/bitcoin/bitcoin/issues/13155 which I believe is the same issue but in libevent.

The issue is on a host that has IPV6 enabled but only a loopback IP address `-proxy=[::1]` will fail as `[::1]` is not considered valid by `getaddrinfo` with `AI_ADDRCONFIG` flag. I think the loopback interface should be considered valid and we have a functional test that will try to test this: `feature_proxy.py`.

To replicate the issue, run `feature_proxy.py` inside a docker container that has IPV6 loopback ::1 address without specifically giving that container an external IPV6 address. This should be the default with recent versions of docker. IPV6 on loopback interface was enabled in docker engine 26 and later ([https://docs.docker.com/engine/release-notes/26.0/#bug-fixes-and-enhancements-2](https://docs.docker.com/engine/release-notes/26.0/#bug-fixes-and-enhancements-2)).

`AI_ADDRCONFIG` was introduced to prevent slow DNS lookups on systems that were IPV4 only.

References:

Man section on `AI_ADDRCONFIG`:

```
If hints.ai_flags includes the AI_ADDRCONFIG flag, then IPv4 addresses are returned in the list pointed to by res only if the local system has at least one IPv4 address configured, and  IPv6  addresses
       are  returned only if the local system has at least one IPv6 address configured.  The loopback address is not considered for this case as valid as a configured address.  This flag is useful on, for ex‐
       ample, IPv4-only systems, to ensure that getaddrinfo() does not return IPv6 socket addresses that would always fail in connect(2) or bind(2).
```

[AI_ADDRCONFIG considered harmful Wiki entry by Fedora](https://fedoraproject.org/wiki/QA/Networking/NameResolution/ADDRCONFIG)

[Mozilla discussing slow DNS without AI_ADDRCONFIG and also localhost issues with it](https://bugzilla.mozilla.org/show_bug.cgi?id=467497)

m3dwards committed:
net: Allow DNS lookups on nodes with IPV6 lo only

AI_ADDRCONFIG prevents ::1 from being considered a valid address on hosts that have a IPV6 loopback IP address but no other IPV6 interfaces.

vasild reviewed (commented):
Concept ACK on allowing usage of `[::1]` on nodes with IPv6 lo only.

A workaround is to use `-dns=0` (which will avoid `AI_ADDRCONFIG` but will make it impossible to use hostnames).

I can reproduce the problem on Linux. FreeBSD treats the loopback address as valid, so this problem is non-existent on FreeBSD.

I think there is a merit in this for our use cases of `getaddrinfo()` / `Lookup*()`:

> IPv4 addresses are returned in the list pointed  to  by res  only if the local system has at least one IPv4 address configured ... _(and same for IPv6)_

But not this:

> The loopback address is  not  considered  for  this case  as valid as a configured address. This flag is useful on, for example, IPv4-only systems, to ensure that getaddrinfo() does not return IPv6 socket addresses that would always fail in connect(2) or bind(2).

This is even bogus to some extent - if the system has only `[::1]` configured and we want to `connect(2)` or `bind(2)` to `[::1]` that will work. So the "always" is too strong / untrue.

---

Now, can we have the `AI_ADDRCONFIG` behavior but get it to consider loopback addresses as valid? For example, after running `getaddrinfo()` with `AI_ADDRCONFIG` run it again without `AI_ADDRCONFIG` and append any loopback addresses from the second run to the results?

m3dwards commented:
> Now, can we have the `AI_ADDRCONFIG` behavior but get it to consider loopback addresses as valid? For example, after running `getaddrinfo()` with `AI_ADDRCONFIG` run it again without `AI_ADDRCONFIG` and append any loopback addresses from the second run to the results?

I like this idea, I'll have a go at implementing it.

m3dwards commented on src/netbase.cpp:
```diff
@@ -56,29 +57,64 @@ std::vector<CNetAddr> WrappedGetAddrInfo(const std::string& name, bool allow_loo
-    const int n_err{getaddrinfo(name.c_str(), nullptr, &ai_hint, &ai_res)};
-    if (n_err != 0) {
-        return {};
-    }
```
I removed the error handling here for two reasons:

Firstly it just returns an empty vector which is what I believe this function will do anyway if it fails to get address info.

Secondly, if `getaddrinfo` doesn't allow loopback only IPV6 then that is returned as an error code 1 "address family for hostname not supported" so this specific case would have to be allowed.

vasild replied:
Are you sure that `getaddrinfo()` would return error (ie `n_err != 0`) and still set something useful in `ai_res`? That sounds strange.

I am worried that it may return an error (for whatever reason) and set `ai_res` to a bogus pointer which we later try to dereference.

Just returning from the lambda and not from `WrappedGetAddrInfo()` if `getaddrinfo()` returns an error seems reasonable to me.

m3dwards replied:
It was working but I agree handling the error is safer.

Also I was wrong that it would need to handle the specific "address family for hostname not supported" error.

Implemented your suggestion of just returning from the lambda.

m3dwards commented:
Pushed new approach of calling `getaddrinfo` twice. First pass is unchanged from original behaviour, second pass adds local IP addresses should they have not been found on the first pass.

vasild commented on src/netbase.cpp:
```diff
@@ -48,36 +48,59 @@ std::vector<CNetAddr> WrappedGetAddrInfo(const std::string& name, bool allow_loo
+    // std::set used over a std::unordered_set as CNetAddr does not have a hash function
+    std::set<CNetAddr> resolved_addresses_set;
+
+    auto callgetaddrinfo = [&](const int flags, const bool only_add_local_addr) {
```
nit: function arguments that are passed by value need not be `const`
```suggestion
    auto callgetaddrinfo = [&](int flags, bool only_add_local_addr) {
```
http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rf-in

vasild commented on src/netbase.cpp:
```diff
@@ -48,36 +48,59 @@ std::vector<CNetAddr> WrappedGetAddrInfo(const std::string& name, bool allow_loo
+                const sockaddr_in6* s6{reinterpret_cast<sockaddr_in6*>(ai_trav->ai_addr)};
+                addr = CNetAddr{s6->sin6_addr, s6->sin6_scope_id};
+            }
+            if (!only_add_local_addr || (only_add_local_addr && addr.IsLocal())) {
```
nit, feel free to ignore if you find the current one more readable, but this is equivalent to:

```suggestion
            if (!only_add_local_addr || addr.IsLocal()) {
```

vasild commented on src/netbase.cpp:
```diff
@@ -48,36 +48,59 @@ std::vector<CNetAddr> WrappedGetAddrInfo(const std::string& name, bool allow_loo
-        }
-        ai_trav = ai_trav->ai_next;
+    resolved_addresses.reserve(resolved_addresses_set.size());
+    for (auto it = resolved_addresses_set.begin(); it != resolved_addresses_set.end(); ) {
+        resolved_addresses.push_back(std::move(resolved_addresses_set.extract(it++).value()));
     }
-    freeaddrinfo(ai_res);
 
     return resolved_addresses;
```
Simpler:

```diff
-    // Convert the set to a vector
-    std::vector<CNetAddr> resolved_addresses;
-    resolved_addresses.reserve(resolved_addresses_set.size());
-    for (auto it = resolved_addresses_set.begin(); it != resolved_addresses_set.end(); ) {
-        resolved_addresses.push_back(std::move(resolved_addresses_set.extract(it++).value()));
-    }
-
-    return resolved_addresses;
+    return {resolved_addresses_set.begin(), resolved_addresses_set.end()};
```

vasild reviewed (commented):
Approach ACK a7cbc27101677ee5ff357da9595f386b03b4756d

m3dwards commented:
@vasild thank you for your review, all comments addressed.

m3dwards commented:
Thanks for your review @vasild, should have addresses all of your comments.

vasild reviewed (approved):
ACK 60a753b7b38fbe89494f8df66f13a84f28af244b

Thank you!

pinheadmz commented on src/netbase.cpp:
```diff
@@ -48,38 +48,57 @@ std::vector<CNetAddr> WrappedGetAddrInfo(const std::string& name, bool allow_loo
+        ai_hint.ai_flags = flags;
+        addrinfo* ai_res{nullptr};
+        const int n_err{getaddrinfo(name.c_str(), nullptr, &ai_hint, &ai_res)};
+        if (n_err != 0) {
+            return;
```
You could check specifically for `EAI_FAMILY` and then only recheck without `AI_ADDRCONFIG` in that condition. I think the code you have now will always check everything twice, but ignore duplicates?

tdb3 replied:
This makes sense to me as well and it might make for a significantly smaller code change.

m3dwards replied:
We still need to check again if we do get the error though so checking the error would be more performant (not always having to call `getaddrinfo` twice) but _more_ code.

m3dwards replied:
I looked into catching specifically `EAI_ADDRFAMILY` errors but this only applies on GNU Libc and is not POSIX. I checked windows and they have different errors. Rather than try to catalogue every possible system and libc implementation, the code now retries on any error (assuming you first tried with `AI_ADDRCONFIG` in the first place).

pinheadmz reviewed (approved):
ACK 60a753b7b38fbe89494f8df66f13a84f28af244b

Tested in an ubuntu docker container with local ipv6 but not external ipv6:

```
# ip address | grep inet 
    inet 127.0.0.1/8 scope host lo
    inet6 ::1/128 scope host 
    inet 172.17.0.2/16 brd 172.17.255.255 scope global eth0
```

Confirmed the `feature_proxy` test fails without the patch. Added extra logging to confirm the error:

```
getaddrinfo: ::1 error: Address family for hostname not supported
```

With the patch, `getaddrinfo` is called a second time with `flags=0` and the call succeeds.

A few questions below.

<details><summary>Show Signature</summary>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

ACK 60a753b7b38fbe89494f8df66f13a84f28af244b
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCAAdFiEE5hdzzW4BBA4vG9eM5+KYS2KJyToFAmaC23IACgkQ5+KYS2KJ
yTrxpQ/8Cz+gj6bv5wNZ/Nuf3N4uV1I2xScdU8LEjt9RfkppplQaDvExq7Bv3rBt
7eSOeuODMWmy+yjSS69NYcuHIg79mDWHNz7XVFX2uqQ0dpix+CaQPYvGDMfiOYR0
XbXNTcx8tCID+TdvPE1QC4iLQitjwwAtD8c9CNupzEy0r5z5JtyxMjUCWcB68LJa
NvKDjRQtO/+D4I7uv0U0Mq+lDFloRYr6byWP9US6Z/mBeBdlvev/82+jkrhstB3q
tYspNR725QBxh5rLjgxoBGjPmjJfyLBpwCGqGFzLVI8BM1oGaa/YYFLE9NK0WRuk
8xKAbXwIu7CkjF859vlgthzROJPq3FHcK/8T9wZFpv6QvDZbaaN4WwpDjI8zFgqI
BH658s+7yNlrjM6JJL5qHD71h3bztvYp6EpzUD1eTqHTY9OumDPc7av4E2ooDDFv
a3xjokiP1JB4qmjqkXExqwdPQ8WayXAOolMAHCu7+IqvhWjzSFBoovbdpChz/kIX
F3fJXbF1qEM8ybjk0Oroja6jVIeCjdKlDajR3ZUmFevkTTMRy7kDPktlbKXKBFz7
uItf5JZ94aQg9TobJLbKXAkTUeFY57rt1Ucxn0vs7zGJQCLVloxOXNhc13FX8iD8
KoUULcryDATRMNxKWieBKMA7RnEOAiJjEZmj47X73pnuxsIL9kc=
=zVvS
-----END PGP SIGNATURE-----
```

pinheadmz's public key is [on keybase](https://keybase.io/pinheadmz/pgp_keys.asc?fingerprint=e61773cd6e01040e2f1bd78ce7e2984b6289c93a)
</details>

tdb3 commented on src/netbase.cpp:
```diff
@@ -48,38 +48,57 @@ std::vector<CNetAddr> WrappedGetAddrInfo(const std::string& name, bool allow_loo
+    // AI_ADDRCONFIG on some systems may exclude loopback only addresses, especially IPV6
+    // We perform a second lookup with ai_flags set to 0 and add local addresses to the set
+    if (ai_flags == AI_ADDRCONFIG) {
+        callgetaddrinfo(0, true);
```
Instead of using 0 as the first argument, might be better to remove only `AI_ADDRCONFIG`, e.g. something like
```c++
callgetaddrinfo(ai_flags & (~AI_ADDRCONFIG), true);
```

m3dwards replied:
`ai_flags` only had `AI_ADDRCONFIG` set so without it it would be 0 anyway right?

tdb3 replied:
Yeah, it would.  The thought here is to increase resilience in the code, so that if ai_flags changes we're not hard-coded to 0.  Seems like the scope of this `if` condition is intended as "if AI_ADDRCONFIG set" rather than "is only AI_ADDRCONFIG set" (which is now updated in f95fb4a1d826765553f3319ee6f1024094954073).  I'm thinking we would want `callgetaddrinfo()` to use similar logic (i.e. keep flags defined except for AI_ADDRCONFIG).

m3dwards replied:
Good suggestion thanks, implemented.

tdb3 reviewed (commented):
Concept ACK
Good find.  Left a couple of comments.

m3dwards commented:
Windows CI fails because the error that I'm looking for `EAI_ADDRFAMILY` doesn't exist in Windows CRT.

Trying to think what to do about it.

m3dwards commented:
@pinheadmz @tdb3 I should have addressed all of your comments.

@vasild the change since your review is rather than always check twice and discard duplicates, we now only check on error. I chose not to look for the specific error as it was not POSIX and was different on different platforms.

tdb3 commented on src/netbase.cpp:
```diff
@@ -48,36 +48,50 @@ std::vector<CNetAddr> WrappedGetAddrInfo(const std::string& name, bool allow_loo
-            assert(ai_trav->ai_addrlen >= sizeof(sockaddr_in));
-            resolved_addresses.emplace_back(reinterpret_cast<sockaddr_in*>(ai_trav->ai_addr)->sin_addr);
+
+    auto callgetaddrinfo = [&](int flags) -> int {
```
Unless I'm overlooking something, might be able to simplify/shorten the change by calling `getaddrinfo()` again when `n_err` is non-zero.  Would avoid the lambda function and might also get rid of the standalone `ai_flags`.

Something like:

```c++
addrinfo* ai_res{nullptr};
const int n_err{getaddrinfo(name.c_str(), nullptr, &ai_hint, &ai_res)};
if (n_err != 0 && (ai_hint.ai_flags & AI_ADDRCONFIG) == AI_ADDRCONFIG) {
    // AI_ADDRCONFIG on some systems may exclude loopback-only addresses
    // If first lookup failed we perform a second lookup without AI_ADDRCONFIG
    ai_hint.ai_flags = (ai_hint.ai_flags & ~AI_ADDRCONFIG);
    const int n_err_retry{getaddrinfo(name.c_str(), nullptr, &ai_hint, &ai_res)};
    if (n_err_retry != 0) {
        return {};
    }
}
```

m3dwards replied:
If pursuing this approach do we think it would be prudent to reset `ai_res` to a nullptr before the second call?

tdb3 replied:
Good thought.  At least from an initial look, what makes sense to me would be checking if it's already null and if not call `freeaddrinfo()` (`getaddrinfo()` handles allocation and `freeaddrinfo()` handles cleanup).  I suspect this could be somewhat overkill (or not happen if unsuccessful `getaddrinfo()` never sets ai_res to non-null), but we don't want a memory leak.  I'm still looking for guarantees in documentation (and we'd need to potentially compare guarantees for different platforms).

As a side note, `freeaddrinfo(NULL)` doesn't crash/segfault (at least on Debian/glibc), so the null check before `freeaddrinfo()` might be redundant (but defensive).

m3dwards replied:
I think it's overkill. In all the examples and documentation I've read, `freeaddrinfo()` is only called when `getaddrinfo()` succeeds.

tdb3 replied:
Probably, but the examples I've seen are all typically small and exit the program if `getaddrinfo()` returns non-zero.

tdb3 reviewed (commented):
I think this is really close.  Left one more comment.

tdb3 reviewed (approved):
ACK 23333b7ed243071c9b4e4f04c727556d8065acbb

Tested with a similar approach to https://github.com/bitcoin/bitcoin/pull/30245#pullrequestreview-2151672225

Created an LXC container with an external and loopback IPv4 address, and with a loopback IPv6 address.

```
root@test30245:~/bitcoin# ip a | grep inet
    inet 127.0.0.1/8 scope host lo
    inet6 ::1/128 scope host noprefixroute 
    inet 192.168.10.81/24 brd 192.168.10.255 scope global eth0
```

Inserted some extra log statments:
```
2024-07-15T12:06:09Z =============First try
2024-07-15T12:06:09Z =============n_err != 0
2024-07-15T12:06:09Z =============Trying a second time
2024-07-15T12:06:09Z =============Second time worked!
```

Appeared to work well.

Also synced approx 8000 blocks on signet with a tor proxy set up to listen on [::1]:9050 (`-proxy=[::1]` with bitcoind).

pinheadmz commented on src/netbase.cpp:
```diff
@@ -59,7 +60,17 @@ std::vector<CNetAddr> WrappedGetAddrInfo(const std::string& name, bool allow_loo
+        if ((ai_hint.ai_flags & AI_ADDRCONFIG) == AI_ADDRCONFIG) {
+            // AI_ADDRCONFIG on some systems may exclude loopback-only addresses
+            // If first lookup failed we perform a second lookup without AI_ADDRCONFIG
+            ai_hint.ai_flags = (ai_hint.ai_flags & ~AI_ADDRCONFIG);
```
If you need to retouch, you might consider using a local `addrinfo` variable for the second call since it serves a specific purpose only inside this if condition. It doesn't matter because it's not like it gets used again later, just something I noticed.

m3dwards replied:
Don't have strong feeling either way but there is an argument that doing it with one variable means any changes to flags passed to the first call get automatically added to the second call which is I think better default behaviour.

pinheadmz reviewed (approved):
ACK 23333b7ed243071c9b4e4f04c727556d8065acbb

Reviewed updated code change and tested in linux VM with local ipv6 only.

<details><summary>Show Signature</summary>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

ACK 23333b7ed243071c9b4e4f04c727556d8065acbb
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCAAdFiEE5hdzzW4BBA4vG9eM5+KYS2KJyToFAmaVfcgACgkQ5+KYS2KJ
yTorBw/7BFoooqdMN+Zb000m/qKsUoMhW36vZBdGTuxt3xkWrRHlggm8ym/fQfen
jt9VcWYl6vmZFnU+nUvJbdN+GLTdf+fDT8vdJdYKjKSxmflEIGLRnf5Pktf2E8lG
rycNQqa6rxcMcnHu0p69VWUU2Ay3vXTV3OOZ/UJFMTLo1twLyljxsSwu6JvMSKgM
4qOqFnqxVeJ1fq3Lw7TtjbwvKweDU8iw2B/YE6kXg711u0qdbSHe3MiPiXKl0sRp
RCkFXuO/q0GICUfW+XHqfs4tDikD+5TRhZ4bBRond7KIdYSdizZzgkNdI1elcI9B
/exMlaLD4cJAEeqovhgNpSByI3rJKQMmzUWS2GrzU78Rsk1zg3a0L+XioM7BMrMm
ClZs82CER6VDNApCv1pEZkdF2AsrbCXDT9GmmpIearoTWAKhR2GP9hccUxKOeZB6
qtb5ep5SvyjM+hIj4IBTM6g0i24SvfdP3/9WcuYj5G1Cl7sWVI1tdXJIDP2hMCij
7J9yC9YznQpvx9626yASNpFq/WRcC4Iug7WdTrSlnIBCQDJ9wMt1uzK6Aj/SdDI+
vKE45GqpYag7MPKnNSobDSe83JWssPy8ZNP1W8FmBV1Vd4yRMdKJea4yrmNYDCwC
CiC31ZufOkWBdb4n+rduarigMQxEOQMm/8nqFCKdl7GFaIcUjPQ=
=wUuO
-----END PGP SIGNATURE-----
```

pinheadmz's public key is [on keybase](https://keybase.io/pinheadmz/pgp_keys.asc?fingerprint=e61773cd6e01040e2f1bd78ce7e2984b6289c93a)
</details>

achow101 commented:
ACK 23333b7ed243071c9b4e4f04c727556d8065acbb
